### PR TITLE
[#8838] cfreactor on py3

### PR DESCRIPTION
**Pull request by hawkowl**
https://twistedmatrix.com/trac/ticket/8838

### CODE REVIEW

File: src/twisted/internet/test/test_process.py [hunk 1]
```diff
@@ -862,6 +862,8 @@ class PTYProcessTestsBuilder(ProcessTestsBuilderBase):
     if platform.isWindows():
         skip = "PTYs are not supported on Windows."
     elif platform.isMacOSX():
+        skip = "PTYs are flaky from a Darwin bug. See #8840."
```
Comment (glyph) [inline src/twisted/internet/test/test_process.py:865]: This might be a bit overly broad.  PTYs should totally be _supported_ on OS X (and they do work, to the extent that the platform works), just there are a few tests which are inadvertently exercising edge-cases around `close()`.

File: src/twisted/internet/cfreactor.py [hunk 2]
```diff
@@ -167,36 +167,41 @@ def _socketCallback(self, cfSocket, callbackType,
...
+                        if rw[_WRITE]:
+                            why = readWriteDescriptor.doWrite()
+            except:
+                why = sys.exc_info()[1]
+                log.err()
+            if why:
+                self._disconnectSelectable(readWriteDescriptor, why, isRead)
+
+        readWriteDescriptor = self._fdmap[fd][2]
+        log.callWithLogger(readWriteDescriptor, _drdw)
```
Comment (glyph) [inline src/twisted/internet/cfreactor.py:203]: Let me see if I can infer the purpose of this change: previously it was calling `why = log.callWithLogger(...)` and expecting `why` to get set to something truthy in the case where there was a problem, so we could notice that problem, and `_disconnectSelectable`.  Now we're allowing the exception to bubble up, always catching it, and setting `why =` properly, so in the case where there's an exception we _actually_ disconnect.

OK, that makes sense.

Comment (glyph) [reply src/twisted/internet/cfreactor.py:203]: Setting `why` "properly" meaning that `callWithLogger` _just returns `None`_ in the case where there's an exception (and it logs it), which is not helpful to this code.

File: src/twisted/internet/cfreactor.py [hunk 3]
```diff
@@ -167,36 +167,41 @@ def _socketCallback(self, cfSocket, callbackType,
...
-                # actually attempt a doRead/doWrite first.  -glyph
-                if isRead:
-                    if rw[_READ]:
-                        why = log.callWithLogger(
-                            readWriteDescriptor, readWriteDescriptor.doRead)
+        def _drdw():
+
+            why = None
+            isRead = False
+            src, skt, readWriteDescriptor, rw = self._fdmap[fd]
```
Comment (glyph) [inline src/twisted/internet/cfreactor.py]: I don't like the fact that `readWriteDescriptor` is bound twice, once inside and once outside.  It's never _re_-bound, so could you just move this tuple-unpacking assignment outside of `_drdw`?

File: src/twisted/internet/cfreactor.py [hunk 4]
```diff
@@ -167,36 +167,41 @@ def _socketCallback(self, cfSocket, callbackType,
...
+                    else:
+                        if rw[_WRITE]:
+                            why = readWriteDescriptor.doWrite()
+            except:
+                why = sys.exc_info()[1]
+                log.err()
+            if why:
+                self._disconnectSelectable(readWriteDescriptor, why, isRead)
+
+        readWriteDescriptor = self._fdmap[fd][2]
```
Comment (glyph) [inline src/twisted/internet/cfreactor.py]: (see comment above about re-binding)

File: src/twisted/internet/udp.py [hunk 5]
```diff
@@ -365,6 +365,7 @@ def connectionLost(self, reason=None):
         """
         log.msg('(UDP Port %s Closed)' % self._realPortNumber)
         self._realPortNumber = None
+        self.maxThroughput = -1
```
Comment (glyph) [inline src/twisted/internet/udp.py:368]: WOOP WOOP NEW PUBLIC ATTRIBUTE, WOOP WOOP

File: src/twisted/trial/_asynctest.py [hunk 6]
```diff
@@ -403,3 +403,26 @@ def stop():
         finally:
             results = None
             running.pop()
+
+
+    def _waitUntilAllDisconnected(self, reactor, protocols):
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: This is just _horrible_!  Also: genius.

However, I think explaining why it's horrible might be a good idea here.  Link to jml's post https://jml.io/pages/how-to-disconnect-in-twisted-really.html about how to do this the "right" way, explain that it's using timed polling, that this is a workaround for legacy test cases, etc.

Also, there's no reason to put this on `TestCase`, especially given that it's not called from here.  `self` isn't even used in the body of the function.

This should maybe go in `proto_helpers`

File: src/twisted/trial/_asynctest.py [hunk 7]
```diff
@@ -403,3 +403,26 @@ def stop():
...
+        """
+        Take a list of disconnecting protocols, callback a Deferred when
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: nope, not protocols, transports

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: Oops I mean the reverse

File: src/twisted/trial/_asynctest.py [hunk 8]
```diff
@@ -403,3 +403,26 @@ def stop():
...
+        they're all done.
+
+        @param: The reactor to use.
+        @param: A list of L{ITransport} or similars.
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: not _transports_, _protocols_

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: also don't say "or similar"; either it's `IProtocol` or it's not

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: also maybe you were a little tired?  `@param ...` what?  which param?

File: src/twisted/trial/_asynctest.py [hunk 9]
```diff
@@ -403,3 +403,26 @@ def stop():
...
             results = None
             running.pop()
+
+
+    def _waitUntilAllDisconnected(self, reactor, protocols):
+        """
+        Take a list of disconnecting protocols, callback a Deferred when
+        they're all done.
+
+        @param: The reactor to use.
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: which param?

File: src/twisted/trial/_asynctest.py [hunk 10]
```diff
@@ -403,3 +403,26 @@ def stop():
...
+        d = defer.Deferred()
+        lc = None
+
+        def _check():
+            if not True in [x.transport.connected for x in protocols]:
+                d.callback(None)
+
+        lc = task.LoopingCall(_check)
+        lc.clock = reactor
+        lc.start(0)
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: `0` strikes me as unduly harsh here.  in the case where this doesn't work immediately I don't see any need to burn 100% CPU in the meanwhile.  start has a flag, `now=True`, which will run it once immediately, and that will take care of the case where everything's already ready.

File: src/twisted/trial/_asynctest.py [hunk 11]
```diff
@@ -403,3 +403,26 @@ def stop():
...
+
+        d.addCallback(lambda _: lc.stop())
```
Comment (glyph) [inline src/twisted/trial/_asynctest.py]: Just an implementation hint here: `lc.start` returns a `Deferred` that fires when it's stopped.  So you can just do `return lc.start(...)` and then in `check` do `lc.stop` which serves the dual purpose of stopping the looping call and firing the `Deferred`; bonus, you don't need to instantiate or manage your own `Deferred`.

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: Managing the Deferred is important, too; when it gets timed out, the canceler needs to be sure to _also_ stop the `LoopingCall`.

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: As it happens, `LoopingCall`'s `Deferred` _doesn't_ do this, but that's its bug, not yours :)

Comment (glyph) [reply src/twisted/trial/_asynctest.py]: specifically this bug https://twistedmatrix.com/trac/ticket/6656

File: src/twisted/web/test/test_webclient.py [hunk 12]
```diff
@@ -348,7 +348,11 @@ def tearDown(self):
             proto.transport.loseConnection()
         if connections:
             msg("Some left-over connections; this test is probably buggy.")
-        return self.port.stopListening()
+        self.port.stopListening()
```
Comment (glyph) [inline src/twisted/web/test/test_webclient.py]: You need to honor this `Deferred` as well; `stopListening` is explicitly allowed to return a `Deferred` (see here https://twistedmatrix.com/documents/16.4.1/api/twisted.internet.interfaces.IListeningPort.html#stopListening)

File: tox.ini [hunk 13]
```diff
@@ -8,7 +8,9 @@
 ; * withcov - run the tests wrapped using the coverage.
 ; * nocov - run the tests directly, without using the coverage wrapper.
 ; * wheel - build the wheel distribution
-; * posix - prepare the tests to be run in a Linux/Unix/OSX environment.
+; * posix - prepare the tests to be run in a Linux/Unix/macOS environment.
```
Comment (glyph) [inline tox.ini:11]: okay this explosion of environments is getting kind of ridicuous.  If we can't use platform specifiers, I'm starting to think we should include a `Makefile` or _something_ that does the dance you need to figure out what tox environments are appropriate to run locally.  I have a shell script that does this right now:

``` bash
#!/bin/bash
cd ~/Projects/Twisted;
coverage erase;
detox $(tox -l | grep withcov-posix |
            while read x;
            do
                echo " -e ${x}";
            done
    ) -e apidocs \
      -e narrativedocs \
      -e topfile \
      -e manifest-checker \
      -e txchecker-travis;
coverage combine;
coverage html;
```

This is not a blocker for this branch but it's something we should do soon.

File: tox.ini [hunk 14]
```diff
@@ -51,10 +53,17 @@ deps =
...
 
-    py27-{alldeps,tests,coverage}-posix: pysqlite
+    py27-{alldeps,tests,coverage}-{posix,macos}: pysqlite
     py27-{alldeps,tests,coverage}: soappy
 
-    alldeps-windows: pypiwin32
+    windows: pypiwin32
+
+    {alldeps,tests}-macos: pyobjc-core
+    {alldeps,tests}-macos: pyobjc-framework-cfnetwork
```
Comment (glyph) [inline tox.ini]: Also not really related to this branch, but it reminds me, we should be similarly discerning in `macosx_platform`, and not depend on all of `pyobjc`.

Comment (glyph) [reply tox.ini]: Also, canonical capitalization here is `"CFNetwork"`, as per https://pypi.python.org/pypi/pyobjc-framework-CFNetwork

File: src/twisted/internet/cfreactor.py (hunk 2 above)
Comment (glyph) [reply src/twisted/internet/cfreactor.py:203]: `callWithLogger` is a bad scene though, for real, we need a new/good logging system replacement for this idiom.

**Review by glyph — COMMENTED** [review]
I wish I could just approve this; these improvements to reliability are pretty awesome!  The implementation shows clear signs of sleep deprivation in places, though, so I think it's probably best if we do one more round of review.

File: src/twisted/internet/udp.py (hunk 5 above)
Comment (hawkowl) [reply src/twisted/internet/udp.py:368]: Existing public attribute!

File: src/twisted/web/test/test_webclient.py (hunk 12 above)
Comment (hawkowl) [reply src/twisted/web/test/test_webclient.py]: I pass self.port to the following function, which should achieve the same effect?

File: src/twisted/internet/test/test_process.py (hunk 1 above)
Comment (hawkowl) [reply src/twisted/internet/test/test_process.py:865]: I had something that shouldn't be related to close crash, though... maybe i need to stress test them individually. Does this happen more on faster PCs than slower? I tried reproducing and it wouldn't...

File: src/twisted/internet/cfreactor.py (hunk 2 above)
Comment (hawkowl) [reply src/twisted/internet/cfreactor.py:203]: What is the idiom? I don't think the idiom exists in the new logging world; loggers are responsible for their own contexts.

Comment (glyph) [inline src/twisted/internet/cfreactor.py:203]: That's why we need one. If it existed, we'd just need to use it.

File: src/twisted/internet/test/test_process.py (hunk 1 above)
Comment (glyph) [reply src/twisted/internet/test/test_process.py:865]: To be clear, we can make re-enabling the PTY tests that _do_ work a separate ticket.  We don't need it here.  Please just file it so that we don't forget.

File: tox.ini (hunk 14 above)
Comment (hawkowl) [reply tox.ini]: Ticket filed on https://twistedmatrix.com/trac/ticket/8848#ticket

File: src/twisted/internet/test/test_process.py (hunk 1 above)
Comment (hawkowl) [reply src/twisted/internet/test/test_process.py:865]: Should #8840 suffice?

File: src/twisted/internet/udp.py (hunk 5 above)
Comment (markrwilliams) [reply src/twisted/internet/udp.py:368]: It's [even documented](https://github.com/twisted/twisted/blob/5c1777777f85aaf799c2269f9f5dddcf0d59803e/src/twisted/internet/udp.py#L66-L67).

File: src/twisted/test/proto_helpers.py [hunk 15]
```diff
@@ -834,3 +834,33 @@ def connectUNIX(self, address, factory, timeout=30, checkPID=0):
...
+    @param reactor: The reactor to use.
+    @type reactor: L{IReactor}
+
+    @param protocols: The protocols to wait for disconnecting.
+    @type protocols: A L{list} of L{IProtocol}s.
+    """
+    lc = None
+
+    def _check():
+        if not True in [x.transport.connected for x in protocols]:
```
Comment (markrwilliams) [inline src/twisted/test/proto_helpers.py:861]: I think this reads better:

``` python
if not any(p.transport.connected for p in protocols):
```

File: src/twisted/test/proto_helpers.py [hunk 16]
```diff
@@ -834,3 +834,33 @@ def connectUNIX(self, address, factory, timeout=30, checkPID=0):
         Fake L{IReactorUNIX.connectUNIX}, that raises L{_connectException}.
         """
         raise self._connectException
+
+
+
+def waitUntilAllDisconnected(reactor, protocols):
```
Comment (markrwilliams) [inline src/twisted/test/proto_helpers.py:840]: Could this be a private API?

**Review by markrwilliams — COMMENTED** [review]
Minor stuff.  Soon to be merged!

File: src/twisted/test/proto_helpers.py (hunk 16 above)
Comment (hawkowl) [reply src/twisted/test/proto_helpers.py:840]: It is private API, it is in `test.proto_helpers` :)

File: src/twisted/test/proto_helpers.py (hunk 15 above)
Comment (hawkowl) [reply src/twisted/test/proto_helpers.py:861]: wow i did not know about `any()`. will fix.

**Comment by hawkowl** [comment]
Builds are green, I'll merge (since @markrwilliams went to bed :) )
